### PR TITLE
enh(host): Allows to create host services according to host templates

### PR DESCRIPTION
**Pull request by callapa**
## Description

Added a mechanism to create host services when a host is added.

**Fixes** # (issue)

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 19.04.x
- [ ] 19.10.x
- [ ] 20.04.x
- [x] 20.10.x (master)

<h2> How this pull request can be tested ? </h2>

Currently, this PR will only be testable with the centeron-autodiscovery module.

## Checklist

- [x] I followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have made corresponding changes to the **documentation**.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: src/Centreon/Domain/Engine/EngineConfiguration.php [hunk 1]
```diff
@@ -0,0 +1,131 @@
+<?php
+
+/*
+ * Copyright 2005 - 2019 Centreon (https://www.centreon.com/)
```
Comment (sc979) [inline src/Centreon/Domain/Engine/EngineConfiguration.php:4]:
```suggestion
 * Copyright 2005 - 2020 Centreon (https://www.centreon.com/)
```

File: src/Centreon/Domain/AccessControlList/Interfaces/AccessControlListRepositoryInterface.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
+<?php
+
+/*
+ * Copyright 2005 - 2019 Centreon (https://www.centreon.com/)
```
Comment (sc979) [inline src/Centreon/Domain/AccessControlList/Interfaces/AccessControlListRepositoryInterface.php:4]:
```suggestion
 * Copyright 2005 - 2020 Centreon (https://www.centreon.com/)
```

File: src/Centreon/Domain/Engine/Interfaces/EngineConfigurationRepositoryInterface.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
+<?php
+
+/*
+ * Copyright 2005 - 2019 Centreon (https://www.centreon.com/)
```
Comment (sc979) [inline src/Centreon/Domain/Engine/Interfaces/EngineConfigurationRepositoryInterface.php:4]:
```suggestion
 * Copyright 2005 - 2020 Centreon (https://www.centreon.com/)
```

File: src/Centreon/Domain/Engine/Interfaces/EngineConfigurationServiceInterface.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
+<?php
+
+/*
+ * Copyright 2005 - 2019 Centreon (https://www.centreon.com/)
```
Comment (sc979) [inline src/Centreon/Domain/Engine/Interfaces/EngineConfigurationServiceInterface.php:4]:
```suggestion
 * Copyright 2005 - 2020 Centreon (https://www.centreon.com/)
```

File: src/Centreon/Domain/HostConfiguration/Host.php [hunk 5]
```diff
@@ -317,24 +321,29 @@ public function getTemplates(): array
...
+     *
+     * @param Host $hostTemplate
      * @return Host
+     * @throws \InvalidArgumentException
      */
-    public function setTemplates(array $templates): Host
+    public function addTemplate(Host $hostTemplate): Host
     {
-        $this->templates = $templates;
+        if ($hostTemplate->getType() !== Host::TYPE_HOST_TEMPLATE) {
```
Comment (sc979) [inline src/Centreon/Domain/HostConfiguration/Host.php:332]: Could you translate the thrown exception too ?

File: src/Centreon/Domain/ServiceConfiguration/ExtendedService.php [hunk 6]
```diff
@@ -0,0 +1,193 @@
+<?php
+
+/*
+ * Copyright 2005 - 2019 Centreon (https://www.centreon.com/)
```
Comment (sc979) [inline src/Centreon/Domain/ServiceConfiguration/ExtendedService.php:4]:
```suggestion
 * Copyright 2005 - 2020 Centreon (https://www.centreon.com/)
```

File: src/Centreon/Domain/ServiceConfiguration/HostTemplateService.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
+<?php
+
+/*
+ * Copyright 2005 - 2019 Centreon (https://www.centreon.com/)
```
Comment (sc979) [inline src/Centreon/Domain/ServiceConfiguration/HostTemplateService.php:4]:
```suggestion
 * Copyright 2005 - 2020 Centreon (https://www.centreon.com/)
```

File: src/Centreon/Domain/ServiceConfiguration/HostTemplateService.php [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    /**
+     * @param Host $hostTemplate
+     * @return HostTemplateService
+     */
+    public function setHostTemplate(Host $hostTemplate): HostTemplateService
+    {
+        if ($hostTemplate->getType() !== Host::TYPE_HOST_TEMPLATE) {
+            throw new \InvalidArgumentException('This host is not a template');
```
Comment (sc979) [inline src/Centreon/Domain/ServiceConfiguration/HostTemplateService.php:59]:
```suggestion
            throw new \InvalidArgumentException(_('This host is not a template'));
```

File: src/Centreon/Domain/ServiceConfiguration/HostTemplateService.php [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+        return $this->hostTemplate;
+    }
+
+    /**
+     * @param Host $hostTemplate
+     * @return HostTemplateService
+     */
+    public function setHostTemplate(Host $hostTemplate): HostTemplateService
+    {
+        if ($hostTemplate->getType() !== Host::TYPE_HOST_TEMPLATE) {
```
Comment (sc979) [inline src/Centreon/Domain/ServiceConfiguration/HostTemplateService.php:58]: Could you add the translation

File: src/Centreon/Domain/ServiceConfiguration/HostTemplateService.php [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        return $this->serviceTemplate;
+    }
+
+    /**
+     * @param Service $serviceTemplate
+     * @return HostTemplateService
+     */
+    public function setServiceTemplate(Service $serviceTemplate): HostTemplateService
+    {
+        if ($serviceTemplate->getServiceType() !== Service::TYPE_TEMPLATE) {
```
Comment (sc979) [inline src/Centreon/Domain/ServiceConfiguration/HostTemplateService.php:79]: Could you add the translation

File: src/Centreon/Domain/ServiceConfiguration/HostTemplateService.php [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+            throw new \InvalidArgumentException('This service is not a template');
```
Comment (sc979) [inline src/Centreon/Domain/ServiceConfiguration/HostTemplateService.php:80]:
```suggestion
            throw new \InvalidArgumentException(_('This service is not a template'));
```

File: src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php [hunk 12]
```diff
@@ -22,33 +22,216 @@
...
+        return $this;
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function applyServices(Host $host): void
+    {
+        if ($host->getId() == null) {
+            throw new ServiceConfigurationException(_('The host id can not be null'));
```
Comment (sc979) [inline src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php:96]:
```suggestion
            throw new ServiceConfigurationException(_('The host id cannot be null'));
```

File: lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 13]
```diff
@@ -15529,3 +15529,12 @@ msgstr "Le server du type Central : '%s'@'%s' ne correspond pas à celui configu
 
 msgid "Missing mandatory parent address, to link the platform : '%s'@'%s'"
 msgstr "Adresse du server parent manquante, pour pouvoir lier la plateforme : '%s'@'%s'"
+
+msgid "Error when searching for services by host"
+msgstr "Erreur lors de la recherche de services par hôte"
+
+msgid "The host id can not be null"
```
Comment (sc979) [inline lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:15536]:
```suggestion
msgid "The host id cannot be null"
```

File: src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php [hunk 14]
```diff
@@ -22,33 +22,216 @@
...
+        if (empty($host->getTemplates())) {
+            return;
+        }
+        /**
+         * To avoid defining a service description with illegal characters,
+         * we retrieve the engine configuration to retrieve the list of these characters.
+         */
+        $engineConfiguration = $this->engineConfigurationService->findEngineConfigurationByHost($host);
+        if ($engineConfiguration === null) {
+            throw new ServiceConfigurationException(_('Impossible to find the Engine configuration'));
```
Comment (sc979) [inline src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php:108]: Was this sentence already translated ?

File: src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php [hunk 15]
```diff
@@ -22,33 +22,216 @@
...
+        }
+
+        $hostTemplateIds = [];
+
+        /**
+         * Find all host templates recursively and copy their id into the given list.
+         *
+         * **We only retrieve templates that are enabled.**
```
Comment (sc979) [inline src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php:116]:
```suggestion
         * **We only retrieve enabled templates.**
```

File: src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php [hunk 16]
```diff
@@ -22,33 +22,216 @@
...
+
+        $servicesToBeCreated = [];
+
+        /**
+         * Then, we set aside the services to be created.
+         * We must not have two services with the same description (alias of the service template).
+         * The priority order is defined by the list of host templates.
+         * We only retrieve the service templates that are activated.
+         */
+        if (!empty($hostTemplateIds)) {
```
Comment (sc979) [inline src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php:185]: the foreach won't loop if $hostTemplateIds is empty. IMHO, this check is useless as there's no else case

File: src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php [hunk 17]
```diff
@@ -22,33 +22,216 @@
...
+                foreach ($serviceTemplates as $serviceTemplate) {
+                    if (!$serviceTemplate->isActivated()) {
+                        continue;
+                    }
+
+                    $illegalObjectNameCharacters = $engineConfiguration->getIllegalObjectNameCharacters();
+
+                    if (
+                        $serviceTemplate->getAlias() !== null
+                        && $illegalObjectNameCharacters !== null
```
Comment (sc979) [inline src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php:198]: If the field is empty, the query will return null : 
image
image
This mean that, the $servicesToBeCreated, will remain empty

File: src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php [hunk 18]
```diff
@@ -22,33 +22,216 @@
...
+
+    /**
+     * @inheritDoc
+     */
+    public function findHostTemplateServices(array $hostTemplateIds): array
+    {
+        try {
+            return $this->serviceRepository->findHostTemplateServices($hostTemplateIds);
+        } catch (\Throwable $ex) {
+            throw new ServiceConfigurationException('', 0, $ex);
```
Comment (sc979) [inline src/Centreon/Domain/ServiceConfiguration/ServiceConfigurationService.php:233]: Shouldn't a message been throwned ?
If, yes, please add the translation too

File: src/Centreon/Infrastructure/HostConfiguration/HostConfigurationRepositoryRDB.php [hunk 19]
```diff
@@ -287,6 +287,71 @@ public function findHost(int $hostId): ?Host
...
+            INNER JOIN `:db`.host 
+                ON host.host_id = template.host_tpl_id
+            ORDER BY `level`, host_host_id, `order`'
+        );
+        $statement = $this->db->prepare($request);
+        $statement->bindValue(':host_id', $host->getId(), \PDO::PARAM_INT);
+        $statement->execute();
+
+        /**
+         * Adds a host template by browsing all templates recursively until you find the parent template.
```
Comment (sc979) [inline src/Centreon/Infrastructure/HostConfiguration/HostConfigurationRepositoryRDB.php:323]:
```suggestion
         * Add a host template by browsing all templates recursively until you find the parent template.
```

File: src/Centreon/Infrastructure/ServiceConfiguration/ServiceConfigurationRepositoryRDB.php [hunk 20]
```diff
@@ -47,6 +59,102 @@ public function __construct(DatabaseConnection $db, SqlRequestParametersTranslat
...
 
+    /**
+     * @inheritDoc
+     */
+    public function addServicesToHost(Host $host, array $servicesToBeCreated): void
+    {
+        $request = $this->translateDbName(
+            'INSERT INTO `:db`.services 
+            (service_template_model_stm_id, command_command_id, service_description, service_alias, service_locked)
+            VALUES (:template_id, :command_id, :description, :alias, :is_locked, :is_activated)'
```
Comment (sc979) [inline src/Centreon/Infrastructure/ServiceConfiguration/ServiceConfigurationRepositoryRDB.php:70]: :is_activated value, is not linked to any column

**Review by sc979 — CHANGES_REQUESTED** [review]

**Review by sc979 — DISMISSED** [review]

**Review by jeremyjaouen — DISMISSED** [review]

File: lang/fr_FR.UTF-8/LC_MESSAGES/messages.po [hunk 21]
```diff
@@ -15535,3 +15535,24 @@ msgstr "Une erreur est survenue lors de la récupération en DB des utilisateurs
...
+msgstr "La méthode publique %s::%s n'a pas été trouvée"
+
+msgid "The public method %s::%s has no parameters"
+msgstr "La méthode publique %s::%s n'a pas de paramètres"
+
+msgid "This host is not a host template"
+msgstr "Cet hôte n'est pas un modèle d'hôte"
+
+msgid "Error when searching for host and related service templates"
+msgstr "Erreur lors de la recherche de modèles d'hôtes et de services connexes"
```
Comment (kduret) [inline lang/fr_FR.UTF-8/LC_MESSAGES/messages.po:15558]:
```suggestion
msgstr "Erreur lors de la recherche d'hôtes et de modèles de service liés"
```

**Review by kduret — DISMISSED** [review]

**Review by loiclau — DISMISSED** [review]

**Review by loiclau — APPROVED** [review]

**Review by kduret — APPROVED** [review]

**Pull request merged**